src/entities/exit/weighted/weightedExit.ts

```diff
@@ -0,0 +1,181 @@
+import { encodeFunctionData } from 'viem';
+import { Token, TokenAmount, WeightedEncoder } from '../../..';
+import { Address } from '../../../types';
+import { BALANCER_VAULT, MAX_UINT256, ZERO_ADDRESS } from '../../../utils';
+import { vaultAbi } from '../../../abi';
+import { parseExitArgs } from '../../utils/parseExitArgs';
+import {
+    BaseExit,
+    BuildOutput,
+    ExitCallInput,
+    ExitInput,
+    ExitKind,
+    ExitQueryResult,
+} from '../types';
+import { getSortedTokens } from '../../utils';
+import { PoolState, AmountsExit } from '../../types';
+import { doQueryExit } from '../../utils/doQueryExit';
+
+export class WeightedExit implements BaseExit {
+    public async query(
+        input: ExitInput,
+        poolState: PoolState,
+    ): Promise<ExitQueryResult> {
+        // TODO - Validate inputs
+
+        const sortedTokens = getSortedTokens(poolState.tokens, input.chainId);
+
+        const amounts = this.getAmountsQuery(sortedTokens, input);
+
+        const userData = this.encodeUserData(input.kind, amounts);
+
+        // tokensOut will have zero address if exit with native asset
+        const { args, tokensOut } = parseExitArgs({
+            chainId: input.chainId,
+            exitWithNativeAsset: !!input.exitWithNativeAsset,
+            poolId: poolState.id,
+            sortedTokens,
+            sender: ZERO_ADDRESS,
+            recipient: ZERO_ADDRESS,
+            minAmountsOut: amounts.minAmountsOut,
+            userData,
+            toInternalBalance: !!input.toInternalBalance,
+        });
+
+        const queryResult = await doQueryExit(
+            input.rpcUrl,
+            input.chainId,
+            args,
+        );
+
+        const bpt = new Token(input.chainId, poolState.address, 18);
+        const bptIn = TokenAmount.fromRawAmount(bpt, queryResult.bptIn);
+
+        const amountsOut = queryResult.amountsOut.map((a, i) =>
+            TokenAmount.fromRawAmount(tokensOut[i], a),
+        );
+
+        return {
+            exitKind: input.kind,
+            id: poolState.id,
+            bptIn,
+            amountsOut,
+            tokenOutIndex: amounts.tokenOutIndex,
+        };
+    }
+
+    private getAmountsQuery(tokens: Token[], input: ExitInput): AmountsExit {
+        switch (input.kind) {
+            case ExitKind.UNBALANCED:
+                return {
+                    minAmountsOut: tokens.map(
+                        (t) =>
+                            input.amountsOut.find((a) => a.token.isEqual(t))
+                                ?.amount ?? 0n,
+                    ),
+                    tokenOutIndex: undefined,
+                    maxBptAmountIn: MAX_UINT256,
+                };
+            case ExitKind.SINGLE_ASSET:
+                return {
+                    minAmountsOut: Array(tokens.length).fill(0n),
+                    tokenOutIndex: tokens.findIndex(
+                        (t) => t.address === input.tokenOut.toLowerCase(),
```

Comment:
Would try to move sanitizing of input outside of the individual implementations. This isn't something we want to remember to do at this level, imo.

Reply:
Thanks for the pointer, added `isSameAddress` to Token implementation.

Reply:
This is fine, but it'd be nicer if sanitize happened before it ever got to `weightedExit`, so we don't even need to worry about this.